### PR TITLE
frame cache caveats documentation

### PR DESCRIPTION
**Pull request by raskle**
## Overview
Added documentation describing the various caveats with the frame cache.  This should explain the TopN behavior in those edge cases

## Pull request checklist

- [x] I have read the [contributing guide](https://github.com/pilosa/pilosa/blob/master/CONTRIBUTING.md).
- [x] I have agreed to the [Contributor License Agreement](https://cla-assistant.io/pilosa/pilosa).
- [x] I have updated the [documentation](https://github.com/pilosa/pilosa/tree/master/docs).
- [x] I have resolved any merge conflicts.
- [x] I have included tests that cover my changes.
- [x] All new and existing tests pass.

## Code review checklist
This is the checklist that the reviewer will follow while reviewing your pull request. You do not need to do anything with this checklist, but be aware of what the reviewer will be looking for.

- [ ] Ensure that any changes to external docs have been included in this pull request.
- [ ] If the changes require that minor/major versions need to be updated, tag the PR appropriately.
- [ ] Ensure the new code is [properly commented](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments) and follows [Idiomatic Go](https://dmitri.shuralyov.com/idiomatic-go).
- [ ] Check that tests have been written and that they cover the new functionality.
- [ ] Run tests and ensure they pass.
- [ ] Build and run the code, performing any applicable integration testing.

### CODE REVIEW

File: docs/query-language.md [hunk 1]
```diff
@@ -341,6 +341,14 @@ have the attribute specified by `field` with one of the values specified in
 
 **Result Type:** array of key/count objects
 
+**Caveats:**
+* TopN on a frame with ranked cache will return the top bitmaps sorted by count.
+* Non-ranked frames maintain a LRU (least recently used) cache, thus a TopN query on this frame will return bitmaps sorted in most recently setbit order.
```
Comment (travisturner) [inline docs/query-language.md]: Frames with cache type `lru` will maintain an LRU (Least Recently Used) cache, thus a `TopN()` query on this type of frame will return bitmaps sorted in order of most recently set bit.

File: docs/query-language.md [hunk 2]
```diff
@@ -341,6 +341,14 @@ have the attribute specified by `field` with one of the values specified in
...
+* The frame's cache size determines the limit of sorted bitmaps.
```
Comment (travisturner) [inline docs/query-language.md]: The frame's cache size determines the number of sorted bitmaps to maintain in the cache for purposes of `TopN()` queries. There is a tradeoff between performance and accuracy; increasing the cache size will improve accuracy of results at the cost of performance.

File: docs/query-language.md [hunk 3]
```diff
@@ -341,6 +341,14 @@ have the attribute specified by `field` with one of the values specified in
 
 **Result Type:** array of key/count objects
 
+**Caveats:**
+* TopN on a frame with ranked cache will return the top bitmaps sorted by count.
+* Non-ranked frames maintain a LRU (least recently used) cache, thus a TopN query on this frame will return bitmaps sorted in most recently setbit order.
+* The frame's cache size determines the limit of sorted bitmaps.  
+* The ranked cache maintains a sorted set of bitmaps by count up to the cache size.
```
Comment (travisturner) [inline docs/query-language.md]: It seems like this bullet point is redundant. Are you trying to convey something specific here?

File: docs/query-language.md [hunk 4]
```diff
@@ -341,6 +341,14 @@ have the attribute specified by `field` with one of the values specified in
 
 **Result Type:** array of key/count objects
 
+**Caveats:**
+* TopN on a frame with ranked cache will return the top bitmaps sorted by count.
+* Non-ranked frames maintain a LRU (least recently used) cache, thus a TopN query on this frame will return bitmaps sorted in most recently setbit order.
+* The frame's cache size determines the limit of sorted bitmaps.  
+* The ranked cache maintains a sorted set of bitmaps by count up to the cache size.  
+* Once full, the cache will truncate the set of bitmaps at the frame options CacheSize limit.  Bitmaps with a similar count that straddle the limit will be truncated in no particular order.
```
Comment (travisturner) [inline docs/query-language.md]: Once full, the cache will truncate the set of bitmaps according to the frame option `CacheSize`.  Bitmaps that straddle the limit and have the same count will be truncated in no particular order.

File: docs/query-language.md [hunk 5]
```diff
@@ -341,6 +341,14 @@ have the attribute specified by `field` with one of the values specified in
...
+* The TopN filter operation only applies to data in the cache, and not all bitmaps in the frame.
```
Comment (travisturner) [inline docs/query-language.md]: I don't understand what this bullet is trying to convey.

Comment (raskle) [inline docs/query-language.md]: I am trying to answer the original question here from #723.  I have clarified my point in the next commit.

File: docs/query-language.md [hunk 6]
```diff
@@ -341,6 +341,13 @@ have the attribute specified by `field` with one of the values specified in
 
 **Result Type:** array of key/count objects
 
+**Caveats:**
+* Performing a TopN() query on a frame with cache type ranked will return the top bitmaps sorted by count in descending order.
+* Frames with cache type lru will maintain an LRU (Least Recently Used) cache, thus a TopN() query on this type of frame will return bitmaps sorted in order of most recently set bit.
+* The frame's cache size determines the number of sorted bitmaps to maintain in the cache for purposes of TopN() queries. There is a tradeoff between performance and accuracy; increasing the cache size will improve accuracy of results at the cost of performance.
+* Once full, the cache will truncate the set of bitmaps according to the frame option CacheSize. Bitmaps that straddle the limit and have the same count will be truncated in no particular order.
+* The TopN() query's attribute filter is applied the the existing sorted cached of bitmaps.  Bitmaps that fall outside of the sorted cache range, but would pass the filter are ignored.
```
Comment (travisturner) [inline docs/query-language.md]: "the the" > "to the"
"cached" > "cache"
single space between sentences.
", but would pass the filter" > ", even if they would normally pass the filter," (note comma after filter)

**Pull request merged**